HSEARCH-3269 Restore support for JSR-352 integration

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3269

Reopen of #2393.

fax4ever commented:
A full build is here: https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2402/4/-

yrodiere commented on batch-jsr352/core/src/main/java/org/hibernate/search/batch/jsr352/core/massindexing/step/lucene/impl/EntityWriter.java:
```diff
@@ -82,6 +83,7 @@ public void open(Serializable checkpoint) throws Exception {
 		searchMapping = Search.mapping( emf );
 		mappingContext = (BatchMappingContext) searchMapping;
 		typeIdentifier = mappingContext.typeContextProvider().typeIdentifierForEntityName( entityName );
+		workspace = searchMapping.scope( typeIdentifier.javaClass(), entityName ).workspace( tenantId );
```
FYI (no need to change it): you can pass `Object.class` instead of `typeIdentifier.javaClass()` here: it's only used to check that entities implement the given class, so we don't really care here.

yrodiere commented on integrationtest/batch-jsr352/src/test/java/org/hibernate/search/integrationtest/batch/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java:
```diff
@@ -11,42 +11,54 @@
 
+	@Rule
+	public OrmSetupHelper ormSetupHelper = OrmSetupHelper
+			.withSingleBackend( new BackendConfiguration() {
+				@Override
+				protected Map<String, Object> backendProperties() {
+					HashMap<String, Object> properties = new HashMap<>();
+					properties.put( "type", "lucene" );
+					properties.put( "type", "local-heap" );
+					properties.put( "multi_tenancy.strategy", "discriminator" );
+					return properties;
+				}
+			} );
```
Just do this, so that indexes are stored on disk, but in the right directory:

```suggestion
			.withSingleBackend( new LuceneBackendConfiguration() );
```

(you may need to add a dependency to `hibernate-search-util-internal-integrationtest-backend-lucene`)


Then add the `multi_tenancy.strategy` option below, in the `setUp` method:

```
.withBackendProperty( "multi_tenancy.strategy", "discriminator" )
```

yrodiere commented on batch-jsr352/core/src/main/java/org/hibernate/search/batch/jsr352/core/massindexing/step/lucene/impl/EntityWriter.java:
```diff
@@ -131,23 +131,40 @@ public void writeItems(List<Object> entities) throws Exception {
 	}
 
-	private void writeItem(PojoIndexer indexer, Object entity) {
+	private void indexAndWaitForCompetitions(List<Object> entities, PojoIndexer indexer) {
```
I think you meant "Completion"?

yrodiere commented on integrationtest/batch-jsr352/src/test/java/org/hibernate/search/integrationtest/batch/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java:
```diff
@@ -70,6 +72,10 @@
 	public void setUp() throws Exception {
+		assumeTrue( "Run this test only when the default persistence unit is in use",
+				PersistenceUnitTestUtil.isDefaultPersistenceUnit()
+		);
```
Alternatively, run the test with Elasticsearch too?

You could rely on the pre-existing `org.hibernate.search.util.impl.integrationtest.common.rule.BackendConfiguration#BACKEND_TYPE` util instead of the custom `org.hibernate.search.jsr352.persistence_unit` property:

* In the failsafe config, set `org.hibernate.search.integrationtest.backend.type` to either `lucene` or `elasticsearch`
* In `PersistenceUnitTestUtil`, infer the PU name from `org.hibernate.search.util.impl.integrationtest.common.rule.BackendConfiguration#BACKEND_TYPE`.
* In `MassIndexingJobWithMultiTenancyIT`, infer the `BackendConfiguration` (either `LuceneBackendConfiguration` or `ElasticsearchBackendConfiguration`) from `org.hibernate.search.util.impl.integrationtest.common.rule.BackendConfiguration#BACKEND_TYPE`. Alternatively, move the if else to a dedicated `BackendConfigurations` class, like we do in the documentation; see `org.hibernate.search.documentation.testsupport.BackendConfigurations`.

yrodiere commented on documentation/src/main/asciidoc/reference/mapper-orm-indexing-jsr352.asciidoc:
```diff
@@ -19,18 +19,16 @@ implementation (see <<jsr-352-emf-jberet,how to configure it here>>).
 ----
+<1> Define a new `MassIndexingJob`.
+<2> Define some parameters. In this case, the list of the entity types to be indexed.
+<3> Get the `JobOperator` form the framework.
```
```suggestion
<3> Get the `JobOperator` from the framework.
```

yrodiere commented on documentation/src/main/asciidoc/reference/mapper-orm-indexing-jsr352.asciidoc:
```diff
@@ -155,23 +153,17 @@ The mass indexing job allows you to define your own entities to be indexed -- yo
-        .build();
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmBatchJsr352IT.java[tags=hql]
 ----
+<1> Define a new `MassIndexingJob`.
```
```suggestion
<1> Start building parameters for a mass-indexing job.
```

yrodiere commented on documentation/src/main/asciidoc/reference/mapper-orm-indexing-jsr352.asciidoc:
```diff
@@ -19,18 +19,16 @@ implementation (see <<jsr-352-emf-jberet,how to configure it here>>).
-        .start( MassIndexingJob.NAME, jobProps );
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmBatchJsr352IT.java[tags=simple]
 ----
+<1> Define a new `MassIndexingJob`.
```
```suggestion
<1> Start building parameters for a mass-indexing job.
```

yrodiere commented on documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmBatchJsr352IT.java:
```diff
@@ -56,13 +56,13 @@ public void setup() {
 		// tag::simple[]
-		Properties jobProps = MassIndexingJob
+		Properties jobProps = MassIndexingJob // <1>
 				.parameters()
```
```suggestion
		Properties jobProps = MassIndexingJob.parameters() // <1>
```

yrodiere replied:
Did you miss this suggestion or decide against applying it?

fax4ever replied:
Actually I thought I applied for the change... Doing now...

yrodiere commented on documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmBatchJsr352IT.java:
```diff
@@ -80,14 +80,14 @@ public void simple() throws Exception {
 		// tag::hql[]
-		Properties jobProps = MassIndexingJob
+		Properties jobProps = MassIndexingJob // <1>
 				.parameters()
```
```suggestion
		Properties jobProps = MassIndexingJob.parameters() // <1>
```

yrodiere replied:
Did you miss this suggestion or decide against applying it?

fax4ever replied:
Actually I thought I applied for the change... Doing now...

yrodiere commented on documentation/src/main/asciidoc/reference/mapper-orm-indexing-jsr352.asciidoc:
```diff
@@ -3,4 +3,473 @@
 // Search 5 anchors backward compatibility
 [[jsr352-integration]]
 
```
I see what you tried to do here; you wanted this file to be considered as "renamed". It didn't work because the file existed before this commit.
Just insert a commit that removes this file, before this commit. In this commit, rename `manual-index.asciidoc` to `mapper-orm-indexing-jsr352.asciidoc`, and change the anchors on the main heading (as above). Then Git should detect the renaming just fine.

yrodiere commented on documentation/src/main/asciidoc/reference/mapper-orm-indexing-jsr352.asciidoc:
```diff
@@ -3,4 +3,473 @@
 
-include::todo-placeholder.asciidoc[]
\ No newline at end of file
+The integration with JSR-352, Batch Applications for the Java Platform, is in development. We do
+think we have the basics covered and we are looking for feedback.
+
```
Ok, I think we can drop this warning, now :)

```suggestion
```

yrodiere reviewed (commented):
Thanks! That was a lot of work :)

I think it's mostly ready to be merged, but here are a few comments.

fax4ever commented on batch-jsr352/core/src/test/java/org/hibernate/search/batch/jsr352/core/massindexing/step/lucene/spi/EntityReaderTest.java:
```diff
@@ -49,14 +47,10 @@
 	);
 
-	@Rule
-	public final MockitoRule mockito = MockitoJUnit.rule().strictness( Strictness.STRICT_STUBS );
```
No reason, it was a mistake. Fixing it..

fax4ever commented on batch-jsr352/core/src/test/java/org/hibernate/search/batch/jsr352/core/massindexing/test/util/JobTestUtil.java:
```diff
@@ -98,14 +93,10 @@ public static JobExecution waitForTermination(JobOperator jobOperator, JobExecut
+		SearchSession searchSession = Search.session( session );
+		return searchSession.search( clazz )
+				.where( f -> f.match().field( key ).matching( value ) )
+				.fetchHits( 1000 );
```
Maybe `.fetch Hits( 1000 )` allows us to avoid the prefetch query.

yrodiere replied:
I'd argue that if there are less than 1000 elements to retrieve, this wouldn't change anything, and if there are more than 1000 elements to retrieve, you probably want to retrieve them all.

Regardless... I don't see how performance optimization is relevant here, but whatever. It's not worth our time, let's leave it like this.

fax4ever commented on batch-jsr352/core/src/test/java/org/hibernate/search/batch/jsr352/core/massindexing/test/entity/Company.java:
```diff
@@ -30,7 +29,7 @@
 	private int id;
 
-	@Field
+	@GenericField
```
This class is merged with the other `Company` of  IT test module.

fax4ever commented on batch-jsr352/core/src/test/java/org/hibernate/search/batch/jsr352/core/massindexing/util/impl/SerializationUtilTest.java:
```diff
@@ -144,12 +137,12 @@ public void deserializeCriteria() throws Exception {
 				.map( Object::toString )
 				.collect( Collectors.toSet() );
-		assertThat( actualCriteriaAsStrings ).containsExactlyInAnyOrder( criterion1.toString(), criterion2.toString() );
+		assertThat( actualCriteriaAsStrings ).containsOnly( criterion1.toString(), criterion2.toString() );
```
I did it when I translated the Hibernate criteria to the JPA criteria. Actually I don't remember why I did it. Anyway criteria API is removed by a subsequent commit. So I don't think it's worth to investigate it if Criteria restriction is no longer supported.

fax4ever commented on batch-jsr352/core/src/main/java/org/hibernate/search/batch/jsr352/massindexing/MassIndexingJob.java:
```diff
@@ -15,14 +15,14 @@
 
 import org.hibernate.CacheMode;
-import org.hibernate.Criteria;
-import org.hibernate.criterion.Criterion;
```
Yes I did and I agree there is no value preserving this commit.
I'll try to remove it, hoping that the rebase will be not very challenging.

fax4ever commented on batch-jsr352/core/pom.xml:
```diff
@@ -68,25 +68,4 @@
         </dependency>
     </dependencies>
-
-    <build>
```
removed in a later commit

yrodiere replied:
`batch-jsr352/core/src/test/resources/META-INF/persistence.xml` is still there.

fax4ever replied:
you right! I didn't remove it

fax4ever commented on batch-jsr352/core/src/main/java/org/hibernate/search/batch/jsr352/core/massindexing/util/impl/JobContextUtil.java:
```diff
@@ -91,9 +94,13 @@ static EntityManagerFactory getEntityManagerFactory(EntityManagerFactoryRegistry
+		Set<Class<?>> entityTypesToIndex = mapping.allIndexedEntities().stream()
+				.map( ie -> ie.javaClass() )
+				.filter( ie -> entityNamesToIndex.contains( ie.getName() ) )
+				.collect( Collectors.toSet() );
```
I did the latter in [this commit](https://github.com/hibernate/hibernate-search/pull/2402/commits/83e0edde2aa3676a1b10d4aeeb6bd095d5ab3375). I don't know if the code is more readable...

fax4ever commented on integrationtest/batch-jsr352/src/test/java/org/hibernate/search/integrationtest/batch/jsr352/massindexing/AbstractBatchIndexingIT.java:
```diff
@@ -100,8 +101,14 @@ protected final void indexSomeCompanies(int count) {
+			for ( Company company : companies ) {
+				indexingPlan.addOrUpdate( company );
+			}
+			indexingPlan.process();
```
Actually not! We have the commit just below that will trigger the process of the plan.

fax4ever replied:
Removing...

fax4ever commented on integrationtest/batch-jsr352/src/test/java/org/hibernate/search/integrationtest/batch/jsr352/massindexing/MassIndexingJobWithCompositeIdIT.java:
```diff
@@ -46,37 +51,36 @@
-		ftem.getTransaction().commit();
-		ftem.close();
+		OrmUtils.withinJPATransaction( emf, ( entityManager ) -> {
+			for ( LocalDate d = START; d.isBefore( END ); d = d.plusDays( 1 ) ) {
```
Sure. It seems that I fabricated it out of thin air... :P

yrodiere commented:
Two other things that I just noticed are missing:

* Adding the new public (non-IT) modules to the distribution ZIP; see https://github.com/hibernate/hibernate-search/commit/ba37e545335c9b8f1e8ea26c412e5e4f076670c1 for an example.
* Reference the new modules (non-IT and IT) in the `reports` module; see https://github.com/hibernate/hibernate-search/commit/a32ee75a454fa44a6a296cea61aed998d789fa25#diff-f8b2caaf2a1ba56b6b7a636cc1be750405191ee4759a4f24a69fac6199b2da05 for an example.

fax4ever commented:
Thank you for the review @yrodiere!
I think I've addressed all the comments with the exception of the new ones: names of parameters and distribution / referencing module.
I'm doing...

fax4ever commented:
Meanwhile I run a new full-build https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2402/17/.
To see if the changes have introduced any regression.

fax4ever commented:
Comments addressed (or at least I hope so). 
Is there anything else we can do it here to improve this contribution?
Thank you

fax4ever commented:
And here is another full-build: https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2402/19.

yrodiere commented:
From what I can see, what you did is not enough to properly include the JSR-352 modules in the distribution ZIP. It's not just about adding a dependency.

Please:

* try again, looking more closely at the commit I gave as reference: https://github.com/hibernate/hibernate-search/commit/ba37e545335c9b8f1e8ea26c412e5e4f076670c1
* and check the result of your work. `mvn clean install -DskipTests -Pdist` should leave a zip file in `distribution/target/zip`, and you can check that the JARs appear there.

fax4ever commented:
Thanks. I checked the zip from distributions, and they seem to be fine.
Basically, the core contains only the package `org.hibernate.search.batch.jsr352.core` and the jberet contains only the package `org.hibernate.search.batch.jsr352.jberet`.
Anything else to do here?

fax4ever commented:
Thanks @yrodiere, I'm waiting the [full CI](https://ci.hibernate.org/blue/organizations/jenkins/hibernate-search/detail/PR-2402/21/pipeline) passes and I'll merge it.

yrodiere commented:
Actually I just noticed the javadoc still doesn't include the JSR-352 module. Investigating...

yrodiere commented on distribution/pom.xml:
```diff
@@ -149,6 +149,7 @@
                                 ${basedir}/../backend/elasticsearch/src/main/java;
                                 ${basedir}/../backend/elasticsearch-aws/src/main/java;
                                 ${basedir}/../backend/lucene/src/main/java;
+                                ${basedir}/../batch-jsr352/core/src/java;
```
```suggestion
                                ${basedir}/../batch-jsr352/core/src/main/java;
```

fax4ever commented:
Thanks, checked. Batch JSR-352 core JavaDocs are there: 
`/hibernate-search/distribution/target/site/apidocs/org/hibernate/search/batch/jsr352/core/*`

fax4ever commented:
Merged thanks!
